src/util/string.h

```diff
@@ -58,34 +58,29 @@ void ReplaceAll(std::string& in_out, const std::string& search, const std::strin
 }
 
 /**
- * Join a list of items
+ * Join items
  *
- * @param list       The list to join
+ * @param collection The items to join
  * @param separator  The separator
- * @param unary_op   Apply this operator to each item in the list
+ * @param unary_op   Apply this operator to each item
  */
-template <typename T, typename BaseType, typename UnaryOp>
-auto Join(const std::vector<T>& list, const BaseType& separator, UnaryOp unary_op)
-    -> decltype(unary_op(list.at(0)))
+template <typename C, typename S, typename UnaryOp>
+auto Join(const C& collection, const S& separator, UnaryOp unary_op)
 {
-    decltype(unary_op(list.at(0))) ret;
-    for (size_t i = 0; i < list.size(); ++i) {
-        if (i > 0) ret += separator;
-        ret += unary_op(list.at(i));
+    decltype(unary_op(*collection.begin())) ret;
+    bool first{true};
+    for (const auto& item : collection) {
+        if (!first) ret += separator;
+        ret += unary_op(item);
+        first = false;
     }
     return ret;
 }
 
-template <typename T, typename T2>
-T Join(const std::vector<T>& list, const T2& separator)
+template <typename C, typename S>
+auto Join(const C& list, const S& separator)
```

Comment:
nit: parameter name inconsistent
```suggestion
auto Join(const C& collection, const S& separator)
```

Reply:
thx, done